lib/leanplum_api/api.rb

```diff
@@ -169,55 +169,51 @@ def development_connection
       @development ||= LeanplumApi::Development.new
     end
 
-    def extract_user_id_or_device_id_hash(hash)
-      user_id = hash['user_id'] || hash[:user_id]
-      device_id = hash['device_id'] || hash[:device_id]
-      fail "No device_id or user_id in hash #{hash}" unless user_id || device_id
+    # Deletes the user_id and device_id key/value pairs from the hash parameter.
+    def extract_user_id_or_device_id_hash!(hash)
+      fail "No device_id or user_id in hash #{hash}" unless hash[:user_id] || hash[:device_id]
 
-      user_id ? { 'userId' => user_id } : { 'deviceId' => device_id }
+      user_id = hash[:user_id] ? { 'userId' => hash[:user_id] } : { 'deviceId' => hash[:device_id] }
+      hash.reject! { |k,v| k.to_s =~ /^(user_id|device_id)$/ }
+      user_id
     end
 
     # Action can be any command that takes a userAttributes param.  "start" (a session) is the other command that most
     # obviously takes userAttributes.
     def build_user_attributes_hash(user_hash, action = 'setUserAttributes')
-      extract_user_id_or_device_id_hash(user_hash).merge(
-        'action' => action,
-        'userAttributes' => turn_date_and_time_values_to_strings(user_hash).reject { |k,v| k.to_s =~ /^(user_id|device_id)$/ }
-      )
+      user_hash = HashWithIndifferentAccess.new(user_hash)
```

Comment:
do you need the hash with indifferent access?
is it because of different (string and symbol) key usages?